### DISCUSSION TITLE
Using ESP32 internal flash partition as an SD card. rebuilding the tools for that UF2, Makecode, SPIFFS, FATFS

### DISCUSSION THREAD

**Discussion by reganissu** [pisto/blastic-scale#6] [Ideas]
UF2, Makecode, SPIFFS, FATFS are the software tools to realize the usage of the Arduino R4's on board ESP32 chip as a simulated usb storage system, to drop from a PC configuration data and retrieve from a PC weight data.

Note that also HW solutions have to be found (if they are possible) to set ESP32 in the right mode to simulate the USB storage when plugged into a PC 

Starting a list of resources to do that, will elaborate with a method and test.

[https://github.com/adafruit/awesome-makecode](https://github.com/adafruit/awesome-makecode)

**Comment by reganissu** [comment]
Finding some material, links below

[https://learn.adafruit.com/adafruit-feather-m4-express-atsamd51/uf2-bootloader-details](https://learn.adafruit.com/adafruit-feather-m4-express-atsamd51/uf2-bootloader-details)

[https://groups.google.com/a/arduino.cc/g/developers/c/arldNYNsFqU/m/GFrBK8rZAAAJ?pli=1](https://groups.google.com/a/arduino.cc/g/developers/c/arldNYNsFqU/m/GFrBK8rZAAAJ?pli=1)

[https://docs.tinyusb.org/en/latest/](https://docs.tinyusb.org/en/latest/)

[https://github.com/arduino/arduino-renesas-bootloader](https://github.com/arduino/arduino-renesas-bootloader)

CHECK CHECK [https://esp32s3.com/install_uf2.html](https://esp32s3.com/install_uf2.html)

**Comment by reganissu** [comment]
[https://github.com/Microsoft/uf2 ](https://github.com/Microsoft/uf2)
"You can also create your own UF2s from binary files using uf2conv.py, available here."

[https://makecode.com/blog/one-chip-to-flash-them-all](https://makecode.com/blog/one-chip-to-flash-them-all)

**Comment by reganissu** [comment]
working on it...
da quanto ho letto si può fare (basta mettere l'esp32 in download mode, shortando i soliti due pin, e sparargli su il bootloader uf2 con esptool) e dalla mia esperienza riuscivo comunque ad uploadare codice da arduino ide, quindi deduco che la port rimane esposta e si può usare anche platformio e il resto.  Quando ho tempo faccio test e se funziona abbiamo sd fat onboard (almeno 1,5 MB ma forse anche di più)
[https://esp32s3.com/install_uf2.html](https://esp32s3.com/install_uf2.html)

**Comment by pisto** [comment]
Not sure about this one. I believe you are mixing up two workflows here:
- configuration persistence (settings are not lost upon reset/power loss)
- data persistence/configuration without serial CLI.

# Configuration persistence

This is already implemented in version v0.2. It uses a simple schema in the Renesas EEPROM. The implementation is not particularly efficient in terms of bytes usage, but it gets away with serialization without too much hassle: the struct is copied as is in EEPROM, and there is a conversion mechanism from older configuration structs to allow for firmware upgrades (and even downgrades in some cases) without losing settings.

It may be debatable whether we should use the Renesas EEPROM or the ESP32 flash for settings. From your research it looks like that you are looking for a way to access the ESP32 flash directly. I argue that accessing the ESP32 in the UNO R4 WiFi is prone to errors, and we risk of conflicting with what the Arduino guys do on the ESP32. We don't really want to ask the user to flash custom stuff on the ESP32, nor we want to spend time to hack the wifi firmware (https://github.com/arduino/uno-r4-wifi-usb-bridge). The user should at most be bothered to install the required official wifi firmware, nothing more.

In any case, you should know that the existing wifi firmware already has an interface to put and retrieve simple settings. As with the wifi data, you use an interface that serializes read/write operations over the serial link, and on the ESP32 side it uses the [non-volatile storage](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/storage/nvs_flash.html). You can see an example [here](https://github.com/arduino/ArduinoCore-renesas/blob/main/libraries/Preferences/examples/Prefs2Struct/Prefs2Struct.ino). IMO this solution does not provide any better value than the current custom implementation in blastic-scale. Flash storage is also finicky (in general you have to work in blocks, you must first erase the block, verify that the erase operation went well, then write a data block aligned to block boundaries, etc. etc.), so I believe it is better to access it directly on the Renesas rather than dealing with the serialization (which has proven to be flawed in some areas).

# Data persistence on SD card

I intend the SD card to be easily swappable between the scale and a PC. CSV data should be recorded on the SD card, and potentially the board may read setttings/commands from a well-known file path.

The filesystem should be readable and writable from all platforms. Unfortunately, the filesystem choice is pretty bleak:
- OS X https://support.apple.com/guide/disk-utility/file-system-formats-dsku19ed921c/mac
- Windows: no official docs? But really it appears that only fat/exfat is supported (+ ntfs but there's no reason to use ntfs, and there is no library for Arduino)
- Linux: not an issue here, supports a large number of filesystems.

The pretty forced choice here is FAT (or exFAT, but we don't really need files larger than 4GB or petabytes of data, so there's really no reason to use it). Am I missing something here?

**Reply by reganissu** [reply]
**Agreeing and moving on**
Thank you for the extended explaination of your doubts, I think that this extended esposition has been useful because of the articulation of the system (code) working in the device, that implies many choices to be made.
I agree on quite some topics, like FAT file system usage choice.  
I agree also on this
> We don't really want to ask the user to flash custom stuff on the ESP32, nor we want to spend time to hack the wifi firmware (https://github.com/arduino/uno-r4-wifi-usb-bridge).

Quite totally agreeing to be true. After considering your answer and trying to build some working solution, I agree with you in the way that this too is a low priority task.  As a consequence we should go on with the SD card implementation and I will work to find the best libraries to choose form for the purpose.
Nevertheless I will write something below about the potential of this solution, because I had it working on other project and enojoyed its possibilities, feel free to skip those consideration, that are here merely for future potential evolution of the device.

**Some consideration for future modification**
I would like to try a simple answer : from my point of view, the idea is the usage of part of the ESP32 flash memory as a sobstitute of an SD card (not a new idea, check for example ["Think of SPIFFS as a very small SD Card onboard the ESP32 chip itself."](https://www.tutorialspoint.com/esp32_for_iot/esp32_for_iot_spiffs_storage.htm), which is referred to spiffs usage, but the same priciple applies to FATfs on ESP32).  It has to be noted that some differences will be however in place opposing to sd card usage:  it will be an usb connection with a cable, much more like and usb stick (with few MB) than an SD card
But why that design choice? 
1) improves the product, making it simpler (less hardware, no need to swap an sd card between a pc or smartphone and the device (adaptor potentially needed, manual dexterity, potential HW failure, workshop enviroment less friendly ecc) ), cheaper, less prone to user error, and better usage of alredy-in-place resources .

2)
>  Flash storage is also finicky [...] I believe it is better to access it directly on the Renesas rather than dealing with the serialization

> I intend the SD card to be easily swappable between the scale and a PC. CSV data should be recorded on the SD card, and potentially the board may read setttings/commands from a well-known file path.

Right. These are the function we can implement 

3) improves the UX, making the scale adoption/usage much more friendly and potentially successful. Reasons:
- drag-and-drop multi-platform (linux, mac ,win) data upload/downlad style
-  simpler physical connection (see point 1)
- OTA (over the air) potential future implementation 
4) The code can be later adapted from an SD card library usage to the internal ESP32 flash usage since they can share the same filesystem type (FAT)
